Experimental load-balancing control-plane with StateDB

joamaki opened the pull request:
This is the beginnings of a new load-balancing control-plane implementation that aims to unify service load-balancing information under a single representation and to allow agent components to observe and update service-related information easily. 

I'm approaching this first with experimental code rather than with a CFP as IMO it's too early to bring up a CFP with wider distribution before there's answers to how to approach the many use-cases. Since there are multiple parties involved in this experiment and it will span a long period of time, I'm proposing to already merge this skeleton implementation to avoid the pain of feature branches. This code is in a separate package, disabled by default and has no production impact other than slightly larger cilium-agent binary.

The rough structure this experiment is working towards is: 
![highlevel](https://github.com/cilium/cilium/assets/1102396/d56ba309-4c60-40d4-851a-30c0329b629f)

Instead of service load-balancing related data being spread over cache.Store, ServiceCache, ServiceManager, etc. we would unify it to 3 sets of tables that would serve the internal information needs and would be used towards BPF map reconciliation. This would solve many of the problems we have with merging service-related data coming from multiple data sources, reduce memory usage and speed up reconciliation throughput as less layers and context switching is involved. See also the [current structure](https://github.com/joamaki/cilium-svc-cfp/blob/main/current.svg) for contrast.

Since we want to mature this implementation gradually and to answer the still many small questions
related to this (topo-aware services, clustermesh, NodePort frontend IP expansion etc.), this PR is
introducing the support via a feature flag (`enable-experimental-services`) so this can be tested and developed
gradually. This hidden feature flag starts a reflector that populates the load-balancing tables and a reconciler
that mocks the reconciliation operations.

Nothing yet uses the new API and the flag is disabled by default, so this code has no impact
to normal operation, but it allows further collaborative development on the API.

The implementation is structured as follows:

`pkg/loadbalancer/experimental` is chosen as the package for these. This clearly separates
the new code from the production code. Eventually it will make sense to move the code into the
`loadbalancer` package that already defines the data structures related to load-balancing.

`pkg/loadbalancer/experimental/{service,frontend,backend}.go`:
These are the load-balancing StateDB tables. To minimize data duplication we're dividing
these into Service (the metadata about the service), Frontend and Backend. To keep things simple
to start with many of fields in Service have been omitted. The Frontend is the object we're reconciling
as that defines a useful unit of work for the BPF map reconciliation as backends need to be updated before frontends and frontends (the services BPF map) directly references them.

`pkg/loadbalancer/experimental/writer.go`:
Implements the "writer" to the load-balancing tables. We don't want to give write access to the tables directly as we want to validate changes and we need to mark frontend status as pending to ask the reconciler to process it.

`pkg/loadbalancer/experimental/reflector.go`:
Observes the Kubernetes Service and EndpointSlice objects and writes to the tables via the `Services` API. Eventually this could be implemented directly with a client-go `Reflector` and skip having a `cache.Store` that holds an unnecessary copy.

`pkg/loadbalancer/experimental/reconciler.go`:
A simulation of what the proper reconciler towards the BPF maps would look like. Just logs the operations we'd need to perform.

The feature flag can be enabled and tested with:
```
$ cilium config set enable-experimental-lb true
# wait for restart

$ kubectl exec -it -n kube-system ds/cilium -- bash
# cilium-dbg statedb experimental frontends
# cilium-dbg statedb experimental backends
# cilium-dbg statedb experimental services
$ kubectl logs -n kube-system ds/cilium | grep loadbalancer-experimental
time=2024-05-28T14:11:12Z level=info msg="Update RevNat" module=agent.controlplane.loadbalancer-experimental.reconciler id=5 address="{AddrCluster:10.96.0.1 L4Addr:{Protocol:TCP Port:443} Scope:0}"
...
```

Next steps following this is to start implementing a test-suite to validate how this API works with all the different use-cases we have for it, and to implement the proper BPF map reconciler for the tables (in progress at https://github.com/joamaki/cilium/pull/32).

joamaki committed:
container: Add JSON marshalling support for ImmSet[T]

For dumping StateDB objects with ImmSet[T] fields, add support
for JSON marshallign and unmarshalling.

The default implementation does not work as ImmSet[T] has
private fields only.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
clustermesh: Add JSON marshalling to AddrCluster

Implement JSON marshalling for AddrCluster type so it can
be used in StateDB objects and dumped with cilium-dbg.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

robscott commented on pkg/loadbalancer/services.go:
```diff
@@ -0,0 +1,415 @@
+		// Do not allow services with same address but different names to override each other.
+		if !existing.Name.Equal(params.Name) {
+			return false, ErrServiceConflict
+		}
```
We can't guarantee the order of operations from Kubernetes API Server so this could be preventing an entirely valid change - ie Service A and B both claim the same IP because the create event for B comes in before the delete event for A. I don't think it's likely, but if we're rejecting this here, we'll need to ensure we have retries with a backoff for anything that's calling UpsertService.

joamaki replied:
Ah that's very interesting. I don't think Cilium currently handles such a possibility. But if it indeed is possible then yes we would need to have retries in the data source implementation.

Or an alternative is to be able to have overlapping services and allow this to happen in any order. E.g. `Shadowed []*Service` field to store the overlapping, but inactive services. That seems fairly complicated so wouldn't go there unless the alternatives are worse.

DamianSawicki replied:
I'd suggest asking ourselves what is a unique and immutable key for Frontends and then querying by this key. Does Frontend correspond to a pair (Service, Port) in Kubernetes? If so, then perhaps we should query by the fully qualified Service name (`namespace+name` in Kubernetes) plus the Port (including protocol).

joamaki replied:
The unique immutable key is the tuple `(address, protocol, port)` (L3n4Addr). The service name is there to group multiple frontends together and to associate the backends with the frontends (it's a secondary index). 

We cannot have the service name in the key since then we may end up with multiple frontends with the same address and that's impossible to reconcile (the assumption is that there's multiple data sources (k8s, REST API, kvstore, ...) writing into this so collisions are possible).

I guess what we could do is allow a frontend to have many names, and thus multiple "backend sets" which we'd merge together. This way multiple data sources could create the same frontend (by IP:port) but call it different things. Not sure if that's more confusing than useful...

robscott commented on pkg/loadbalancer/services.go:
```diff
@@ -0,0 +1,415 @@
+		// another source. If the service is 'Unspec'
+		if existing.Source != params.Source {
+			return false, ErrServiceSourceMismatch
+		}
```
This does seem like a reasonable approach, but any reason not to just overwrite here instead? Would it be worth having some kind of configurable prioritization of Sources where data from Source A always has precedence from data from Source B? What does "Unspec" mean in this context?

joamaki replied:
Mostly I was trying to start out with some reasonable, but strict defaults. I have not fully thought through yet how sources that have overlapping services should interact as it's not really something we've supported before. 

Not sure why I have an unfinished comment here nor what I was thinking about with "Unspec" sources. Will clean it up.

robscott replied:
Thinking about this more, at least in GKE's use case, some form of prioritization, ideally configurable would be ideal. There are some cases where the same Service may transition from one data source to another, and in that case, it would be ideal if we could have a Service present in both data sources to ensure a smooth transition. Essentially that would mean if the Service came from a higher priority data source, we'd transition to that as the source of truth, but if it then disappeared from that data source, we'd transition to the next highest priority data source for that Service.

joamaki replied:
Yeah that's kind of what we have in `pkg/source`. Unfortunately the "AllowOverwrite" is hard-coded, so it's not very useful here. Perhaps the sources should be extendable (e.g. outside Cilium OSS code if one embeds the AgentCell into a derivative project) and they should have an associated priority...

joamaki replied:
Hmm could the priorities be different for different entities?

DamianSawicki replied:
> Hmm could the priorities be different for different entities?

I'm not sure what you mean by entities and what Rob has in mind, but I suppose that it should suffice if it is possible to customise the priorities with some config file (that is, without the need to alter the code).

robscott commented on pkg/loadbalancer/services.go:
```diff
@@ -0,0 +1,415 @@
+}
+
+func (s *Services) deleteService(txn ServiceWriteTxn, svc *Service) error {
+	// Release references to the backends
```
Nit: Maybe make it clear that this will also result in backend deletion
```suggestion
	// Release references to the backends and delete them if they're orphaned
```

robscott commented on pkg/loadbalancer/services.go:
```diff
@@ -0,0 +1,415 @@
+	return err
+}
+
+func (s *Services) DeleteServicesBySource(txn ServiceWriteTxn, source source.Source) error {
```
When do you expect this would be used? Do you expect that it will be possible to add or remove sources while Cilium is actively running?

joamaki replied:
It's needed when a source needs to resync. Plan is to use the client-go Reflector directly without an Informer in which case we need to implement the Replace() operation and that's where we'd use this.

joamaki replied:
https://github.com/cilium/cilium/blob/8f9649a00c94da80cb352487d56578e4673efd75/daemon/tables/k8s_service_reflector.go#L244

robscott commented on pkg/loadbalancer/services.go:
```diff
@@ -0,0 +1,415 @@
+	return nil
+}
+
+func (s *Services) UpsertBackends(txn ServiceWriteTxn, serviceName ServiceName, source source.Source, bes ...*Backend) error {
```
I'm trying to think through how we would translate this from Kubernetes informers and/or as an xDS client. 

It seems like we'd need to do something like this:

1. Get existing set of endpoints in Cilium state with `loadbalancer.GetBackendsForService()`
2. Compare that list with the set of endpoints you've received from Kubernetes (likely by merging the EndpointSlices you've received from k8s API Server) to determine which endpoints need to be added, updated, or deleted
3. Creates go to `UpsertBackends()`, updates go to `UpdateBackendState()`, and deletes go to multiple `ReleaseBackend()` calls

All of the above then needs to be wrapped in some form of transaction where you're blocking writes to endpoints tied to that Service for the duration of those steps. Is that roughly the correct intended usage?

joamaki replied:
I would probably instead hold the state inside the Kubernetes or other data source to figure out from the endpoint update what the delta is. E.g. https://github.com/cilium/cilium/blob/8f9649a00c94da80cb352487d56578e4673efd75/daemon/tables/k8s_service_reflector.go#L96.

robscott commented on pkg/loadbalancer/services.go:
```diff
@@ -0,0 +1,415 @@
+	}
+
+	iter, _ := s.svcs.Get(txn, ServiceNameIndex.Query(serviceName))
+	for svc, _, ok := iter.Next(); ok; svc, _, ok = iter.Next() {
```
I'm probably missing something very obvious here, but why do we need to iterate here? I'm assuming a call to upsertBackends should refer to exactly one Service?

joamaki replied:
No because `Service` in this context really only describes a single frontend. There can be many of them sharing the same name, but having different type, IP or port. Same as how it works currently with `loadbalancer.SVC` and ServiceManager. A case could be made to instead call these things frontends...

joamaki replied:
I've now renamed `Service` to `Frontend`.

robscott commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -341,12 +342,39 @@ type ServiceID uint16
 // ServiceName represents the fully-qualified reference to the service by name,
 // including both the namespace and name of the service (and optionally the cluster).
+// +k8s:deepcopy-gen=true
 type ServiceName struct {
```
This is probably not worth changing given how widely it's already used, but I think a name like `ServiceIdentity` would more clearly describe what this is used for. The `ServiceName` references throughout keep on throwing me off.

joamaki replied:
Yeah, should've called it `ServiceID` instead when I added it.

robscott commented on pkg/loadbalancer/services.go:
```diff
@@ -0,0 +1,415 @@
+			Source:  source,
+		}
+		if old, _, ok := s.bes.First(txn, BackendAddrIndex.Query(be.L3n4Addr)); ok {
+			// UpsertBackends cannot overwrite the backend state or allocated ID.
```
I'm probably just looking at the world through Kubernetes-tinted glasses, but wouldn't you usually have endpoint state if you had the other endpoint info? Why not allow state to be updated here instead of requiring a follow up call to UpsertBackendState?

joamaki replied:
Because we might have the same backend added from multiple sources or from multiple EndpointSlices and we want to retain the learned state/health of the backend even when it's being added multiple times. Though now that I look at this again, we do want to propagate the terminating state from K8s. This definitely needs a rethink.

robscott commented on pkg/loadbalancer/services.go:
```diff
@@ -0,0 +1,415 @@
+
+	params := ServiceParams{
+		Name:                      svc.Name,
+		Frontend:                  svc.Frontend.L3n4Addr,
```
How do you handle multiple addresses/IP families? Are these just considered different Services in this context? Are there separate "Services" for every port as well?

joamaki replied:
Yes, there's a separate service for each type, port, address. I'm starting to lean towards calling this a `Frontend` instead of `Service` as it's confusing when this is not semantically the same as the K8s one.

joamaki replied:
Renamed `Service` to `Frontend`

robscott commented on pkg/service/new_reconciler.go:
```diff
@@ -0,0 +1,152 @@
+		statedb.Filter(iter, func(be *loadbalancer.ServiceBackend) bool {
+			return svc.Frontend.ProtoEqual(&be.L3n4Addr)
+		}))
+	_, _, err := ops.Manager.UpsertService(toSVC(svc.Frontend, svc, bes))
```
How are stale endpoints pruned? Not seeing any evidence of that happening here, but I'm probably just missing something obvious.

joamaki replied:
`ServiceManager` keeps track of backends that are referenced by services and removes the ones that do not exist. The UpsertService call here tells ServiceManager about all the backends currently associated to a service.

I would like to replace `ServiceManager` later on with a StateDB reconciler that takes the `Service` and queries for associated backends and similarly figures out which backends are now gone and need to be removed from the backends BPF map.

Looks roughly like this: https://github.com/cilium/cilium/pull/31755/files#diff-1aab408d0c33a6dfa08997f13c3a94f372eb95a331b731c09752240a960be346R45

robscott commented on pkg/loadbalancer/services.go:
```diff
@@ -0,0 +1,415 @@
+	// Release references to the backends
+	iter, _ := s.bes.Get(txn, BackendServiceIndex.Query(svc.Name))
+	for be, _, ok := iter.Next(); ok; be, _, ok = iter.Next() {
+		be, orphan := be.removeRef(svc.Name)
```
It seems at least theoretically possible that some race conditions or errors here could leave orphaned backends around indefinitely. I'm not really sure what the solution is here, but have you already been thinking about how to handle this?

joamaki replied:
The idea is that all writes to the tables go via the Services API that will take care of managing the references. The solution is to implement that correctly :-)

robscott reviewed (commented):
Thanks for all the work on this @joamaki! This looks really promising, mostly just have a bunch of questions to make sure I'm understanding things correctly.

DamianSawicki commented on pkg/loadbalancer/experimental/tables.go:
```diff
@@ -0,0 +1,299 @@
+type FrontendParams struct {
+	Name     loadbalancer.ServiceName // Fully qualified service name
+	Source   source.Source
+	Frontend loadbalancer.L3n4Addr // Frontend address
```
Struct `Frontend` has a field of type `FrontendParams`, which in turn has a field called `Frontend` again yet of a different type. I'd suggest renaming the last one to something like `Address`, which would eliminate the need for a comment in this line.

joamaki replied:
Ah good point, will rename.

joamaki commented on pkg/loadbalancer/experimental/services.go:
```diff
@@ -0,0 +1,394 @@
+
+	existing, _, found := s.fes.Get(txn, FrontendL3n4AddrIndex.Query(params.Address))
+	if found {
+		// Do not allow frontends with the same address but different names to override each other.
```
I'm a bit unsure about these validations. Should we consider just dropping them at this point as we're still figuring out what semantics make sense?

DamianSawicki replied:
Yeah, good idea!

DamianSawicki commented on pkg/loadbalancer/experimental/benchmark_test.go:
```diff
@@ -0,0 +1,222 @@
+	}
+	wtxn.Commit()
+
+	// Benchmark the speed at which a new service is upserted. 'numObjects' are inserted in one
```
Nit: The comment is a bit confusing: large `numObjects` amortises `p.Services.WriteTxn()` from line 46 and `wtxn.Abort()` from line 62 (Abort, not Commit). However, `b.N` should indeed amortise `p.Services.WriteTxn()` and `wtxn.Commit()` from lines 25 and 41, but this is not needed as we can `b.ResetTimer()` instead as in line 164.

joamaki replied:
Oh yeah indeed confusing. Fixed. The benchmark itself is doing what's intended: against 1k existing services & frontends, insert 1k new ones repeatedly (throwing away the result, so repeating the same operation over and over).
EDIT: Split the benchmark into two, one for txn's of size 1, and one for txn's of size 100 (beyond this size there's not much perf. benefit in StateDB).

DamianSawicki commented on pkg/loadbalancer/experimental/benchmark_test.go:
```diff
@@ -0,0 +1,222 @@
+			},
+		)
+	}
+	wtxn.Abort()
```
According to the comment from line 92, these are 'existing' objects, so I'd assume the transaction should be committed.

joamaki replied:
Yep!

DamianSawicki commented on pkg/loadbalancer/experimental/benchmark_test.go:
```diff
@@ -0,0 +1,222 @@
+func BenchmarkReplaceBackend(b *testing.B) {
+	p := servicesFixture(b)
+
+	b.ResetTimer()
```
This is not needed as we reset the timer in line 164.

DamianSawicki commented on pkg/loadbalancer/experimental/benchmark_test.go:
```diff
@@ -0,0 +1,222 @@
+func BenchmarkReplaceService(b *testing.B) {
+	p := servicesFixture(b)
+
+	b.ResetTimer()
```
The timer is reset in line 203, so this is effectively a NoOp and can be removed.

DamianSawicki commented on pkg/loadbalancer/experimental/reconciler.go:
```diff
@@ -0,0 +1,134 @@
+
+func toBackends(bes []*Backend) []*loadbalancer.Backend {
+	out := make([]*loadbalancer.Backend, len(bes))
+	for i := range bes {
```
Nit: it would probably be more golang-idiomatic to write `i, b := range bes` and then use `b` instead of `bes[i]`.

DamianSawicki commented on pkg/loadbalancer/experimental/services.go:
```diff
@@ -0,0 +1,399 @@
+	return
+}
+
+func (s *Services) DeleteFrontendsByName(txn ServiceWriteTxn, name loadbalancer.ServiceName, source source.Source) error {
```
rename `Frontends` -> `Frontend`

DamianSawicki commented on pkg/loadbalancer/experimental/services.go:
```diff
@@ -0,0 +1,399 @@
+			// Keep the old state, e.g. ID
+			be = *old
+		} else {
+			be.ReferencedBy = NewServiceNameSet(serviceName)
```
We insert `serviceName` below, so here the argument list can probably be empty.

DamianSawicki commented on pkg/loadbalancer/experimental/services.go:
```diff
@@ -0,0 +1,399 @@
+	// reconciliation.
+	revision := s.bes.Revision(txn)
+	for name := range names {
+		fe, _, found := s.fes.Get(txn, FrontendNameIndex.Query(name))
```
In line 335 we use the `List` method for this index, and the index has `Unique: false` in `pkg/loadbalancer/experimental/tables.go`, line 253, so I'd suppose we should use `List` instead of `Get` also here.

DamianSawicki commented on pkg/loadbalancer/experimental/services.go:
```diff
@@ -0,0 +1,399 @@
+	// Bump the backend revision for each of the referenced services to force
+	// reconciliation.
+	revision := s.bes.Revision(txn)
+	iter := s.fes.List(txn, FrontendNameIndex.Query(name))
```
This line and the following loop occur in identical form in `ReleaseBackend` and `UpdateBackendState`, and they should probably also occur in `DeleteBackendsBySource` (see my comment there). I'd suggest to isolate them to a helper function.

DamianSawicki commented on pkg/loadbalancer/experimental/services_test.go:
```diff
@@ -0,0 +1,411 @@
+
+	// Lookup
+	{
+		txn := p.DB.ReadTxn()
```
Please correct me if I'm missing something here. It seems more consistent to expose `ReadTxn` in `Services` (similarly to how we expose `func (s *Services) Frontends() statedb.Table[*Frontend]` for read-only access) so that we can write here `txn := p.Services.ReadTxn()` analogous to `wtxn := p.Services.WriteTxn()` in line 82 and 119.

joamaki replied:
Agreed. Though do wonder whether it'd be cleaner to not have these "convenience methods" at all.

DamianSawicki commented on pkg/loadbalancer/experimental/services_test.go:
```diff
@@ -0,0 +1,411 @@
+		svc, _, found := p.FrontendTable.Get(txn, experimental.FrontendL3n4AddrIndex.Query(frontend))
+		if assert.True(t, found, "Service not found by addr") {
+			assert.NotNil(t, svc)
+			assert.Equal(t, name, svc.Name, "Service name not equal")
```
In some other places, for both `require.` and `assert.`, the string describes the desired state, e.g. `"Service created"` in line 93 or `"Marked pending"` in line 106, and here it describes the undesired state `"Service name not equal"`. It would be nice to have it all consistent.

DamianSawicki commented on pkg/loadbalancer/experimental/services_test.go:
```diff
@@ -0,0 +1,411 @@
+		_, _, found := p.FrontendTable.Get(wtxn, experimental.FrontendNameIndex.Query(name))
+		assert.False(t, found, "Frontend found after delete")
+
+		wtxn.Abort()
```
Don't we want to `Commit` at least one of these `wtxn` transactions and perform a fresh lookup?

joamaki replied:
No this is on purpose forgetting the deletion so the next test case can again access these.

DamianSawicki replied:
This is the last line of the function, so there is no "next test case", right?

joamaki replied:
Sure, but I'd still leave it there so if we ever add more into this we don't end up testing against unexpected state.

DamianSawicki commented on pkg/loadbalancer/experimental/services_test.go:
```diff
@@ -0,0 +1,411 @@
+		require.True(t, bes[0].L3n4Addr.DeepEqual(&beAddr3))
+	}
+
+	// GetBackendsForService
```
Rename to `GetBackendsForFrontend`?

DamianSawicki commented on pkg/loadbalancer/experimental/tables.go:
```diff
@@ -0,0 +1,282 @@
+
+// These tables along with the [Services] API implement a new experimental internal API for
+// configuring service load-balancing. The tables can be inspected with the hidden cilium-dbg
+// commands "cilium-dbg statedb services" and "cilium-dbg statedb backends".
```
According to `Documentation/cmdref/cilium-dbg_statedb_experimental.md`, the command is
```
cilium-dbg statedb experimental frontends
```
and not
```
cilium-dbg statedb services
```

Similarly, `Service` and `ServiceParams` below should probably be `Frontend` and `FrontendParams` respectively.

DamianSawicki commented on pkg/loadbalancer/experimental/tables.go:
```diff
@@ -0,0 +1,282 @@
+//
+// This embeds the Backend and adds further fields to it. Eventually
+// we'd unify them, but for now we do not want to affect existing code yet.
+// We would rename Backend to Backend and existing Backend to BackendParams
```
`rename Backend to Backend`? This looks like a typo.

DamianSawicki commented on pkg/loadbalancer/experimental/tables.go:
```diff
@@ -0,0 +1,282 @@
+
+// Backend describes a backend for one or more services.
+//
+// This embeds the Backend and adds further fields to it. Eventually
```
Nit: I'd suggest adding `loadbalancer.` before `Backend` for clarity.

DamianSawicki commented on pkg/loadbalancer/experimental/tables.go:
```diff
@@ -0,0 +1,282 @@
+	LoopbackHostport          bool
+}
+
+// Clone returns a shallow clone of ServiceParams.
```
`ServiceParams` -> `FrontendParams`

DamianSawicki commented on pkg/loadbalancer/experimental/tables.go:
```diff
@@ -0,0 +1,282 @@
+	// of the frontend (rest of the fields are managed by Services).
+	// This is embedded by reference to avoid copying when updating just the other fields.
+	//
+	// This is immutable! If you want to update the frontend you must Clone() it first!
```
`This is immutable! If you want to update the frontend you must Clone() it first!`
->
`This is immutable! If you want to update the FrontendParams you must Clone() it first!`

DamianSawicki commented on pkg/loadbalancer/experimental/tables.go:
```diff
@@ -0,0 +1,282 @@
+	// maps to refer to this service.
+	ID loadbalancer.ID
+
+	// BackendRevision is the backends table revision of the latest updated backend
```
The first sentence of the comment is not technically true because `BackendRevision` is also bumped e.g. in `ReleaseBackend`. After releasing a backend, `BackendRevision` is more recent than the revision of any of the backends referencing this service. Perhaps we can add "Roughly speaking" before this sentence?

DamianSawicki commented on pkg/loadbalancer/experimental/tables.go:
```diff
@@ -0,0 +1,282 @@
+	}
+}
+
+// Clone returns a shallow copy of the service.
```
service -> frontend

DamianSawicki commented on pkg/loadbalancer/experimental/wrapper.go:
```diff
@@ -0,0 +1,85 @@
+
+	wtxn.Commit()
+
+	return created, 0, nil
```
If this wrapper always returns `0` as `loadbalancer.ID`, it's not a true replacement of a ServiceManager, right?

DamianSawicki commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -789,6 +815,13 @@ func (a *L3n4Addr) IsIPv6() bool {
 }
 
+// ProtoEqual returns true if protocols match for both L3 and L4.
+func (l *L3n4Addr) ProtoEqual(o *L3n4Addr) bool {
```
Since this tests both L3 and L4, perhaps the name `ProtocolsEqual` (note the plural Protocol**s**) would be more self-explanatory. Also, "Proto" is a common abbreviation for "Protocol Buffer", so personally I'd not avoid abbreviating here.

DamianSawicki reviewed (commented):
Not very material this time: one question and the rest are just typos or rewording suggestions.

joamaki committed:
experimental: Add fuzz test to validate serializability

The "cilium-dbg statedb" commands require the objects to be
JSON serializable, which is easy to break accidentally.

Add a fuzz test to generate arbitrary Service, Frontend and
Backend objects and validate that the TableRow() output is the
same across JSON serialization.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer/experimental: Add experimental API for load-balancing

Add experimental Services API for managing load-balancing frontends
and backends. This is added as a new experimental package to avoid
confusing it with the production implementation.

When the hidden "--enable-experimental-services" flag is set:
* K8s Service and Endpoints are reflected to service, frontend and backend tables
* A mock reconciler is started that logs mock operations to reconcile the frontends.

The tables can be inspected with "cilium-dbg statedb experimental" commands,
e.g. "cilium-dbg statedb experimental frontends".

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb commented on pkg/loadbalancer/experimental/cell.go:
```diff
@@ -0,0 +1,52 @@
+// ServicesCell provides the [Services] API for configuring load-balancing and the
+// Table[*Service], Table[*Frontend] and Table[*Backend] for read-only access
+// to load-balancing state.
+var ServicesCell = cell.Module(
```
Maybe ServicesHandlingCell? Otherwise, it's easy to mix with a svc instance (same problem exists today in pkg/service).

joamaki replied:
How about calling `Services` the `Writer`? This much better reflects what it is about. Later it could be used as `loadbalancer.Writer`.

joestringer commented on pkg/loadbalancer/experimental/reconciler.go:
```diff
@@ -0,0 +1,244 @@
+	// Not sure if it makes a difference.
+
+	sort.Slice(bes, func(i, j int) bool {
+		// sorting order: state, preferred, address, port
```
What does `preferred` refer to here?

joamaki replied:
Oops, leftover. Cleaned.

joestringer commented on pkg/loadbalancer/experimental/reconciler.go:
```diff
@@ -0,0 +1,244 @@
+	for be, rev, ok := iter.Next(); ok; be, rev, ok = iter.Next() {
+		backendRevisions[be] = rev
+	}
+	orderedBackends := sortedBackends(maps.Keys(backendRevisions))
```
What property of the sorting is important here? Is it primarily that you want to ensure that Active services get updated before terminating, quarantined, etc? Or is it just that making updates deterministic would be easier to reason about, debug, etc?

joamaki replied:
Sorting in general is important to retain the same order in the BPF maps. As long as the order is deterministic the specific properties are not important, though definitely it needs to follow prior implementation. I'll document this.

joestringer commented on pkg/loadbalancer/experimental/reconciler.go:
```diff
@@ -0,0 +1,244 @@
+
+func sortedBackends(bes []*Backend) []*Backend {
+	// TODO: this is different from the existing implementation that sorts by ID.
+	// Not sure if it makes a difference.
```
Assuming I'm following the code correctly, commit 27d82b5ab006a9fa38f8beffcc593a0f491072a1 seems to hint that it's just about reflecting reliable results to integration tests. :shrug:

joamaki replied:
The ID sorting will be added later once the implementation allocates the backend IDs (working on that in the follow-up that is https://github.com/joamaki/cilium/pull/32).

joestringer commented on pkg/loadbalancer/experimental/reconciler.go:
```diff
@@ -0,0 +1,244 @@
+	for _, be := range orderedBackends {
+		backendAddrs.Insert(be.L3n4Addr)
+	}
+	for _, orphan := range ops.backendsState.updateReferences(fe.Address, backendAddrs) {
```
nit: If you pushed the set creation into `updateReferences()`, then you wouldn't need to clone the newly created set inside that function just after creating it here. `backendAddrs` doesn't seem to be used for the rest of the function.

joamaki replied:
Cleaned up a bit by moving the set creation there and improved the field names of backendsState. Still need the clone to put into `references` as the set is mutated later.

joestringer commented on pkg/loadbalancer/experimental/services_test.go:
```diff
@@ -0,0 +1,302 @@
+		cfg.EnableExperimentalServices = true
+	})
+
+	require.NoError(t, h.Start(log, context.TODO()))
+	t.Cleanup(func() {
+		h.Stop(log, context.TODO())
+	})
```
I guess the plan would be to resolve all TODOs before this implementation graduates experimental?

joestringer replied:
Related: Same for `panic()` (as called out elsewhere in the commit).

joestringer commented on pkg/loadbalancer/experimental/reconciler.go:
```diff
@@ -0,0 +1,244 @@
+				//log.Infof("Backend %s is now an orphan", addr.StringWithProtocol())
+				orphans = append(orphans, addr)
+			} else {
+				//log.Infof("Backend %s ref count now %d", addr.StringWithProtocol(), count)
```
nit: We'd usually drop inactive code lines like this.
```suggestion
				orphans = append(orphans, addr)
			} else {
```

joestringer commented on pkg/loadbalancer/experimental/reconciler.go:
```diff
@@ -0,0 +1,244 @@
+	// Update RevNat
+	ops.log.Info("Update RevNat", "id", fe.ID, "address", fe.Address)
+
+	// Update the master service and backend slots
```
What's a `master service`? I don't see this terminology elsewhere in the parent package or in this commit.

joamaki replied:
The mock reconciler is essentially replicating the operations performed in the lbmap package and that's where this terminology is from: https://github.com/cilium/cilium/blob/main/pkg/maps/lbmap/lbmap.go#L131. 

I have the proper reconciler partially implemented in https://github.com/joamaki/cilium/pull/32 where this is more evident.

The "master service" is the "slot 0" of the entries in the services BPF map, e.g. the one we look up first to get the backend count after which the backend can be looked up (slots 1..N).

joestringer replied:
OK fair enough. We may want to consider referring to this as the "primary key" or such.

joestringer commented on pkg/loadbalancer/experimental/reconciler.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+// Update implements reconciler.Operations.
+func (ops *mockOps) Update(ctx context.Context, txn statedb.ReadTxn, fe *Frontend) error {
```
At a high level, I feel like scaling up / down backends is a common place for LB to introduce undesirable behaviour, such as distorting the distribution to backends. Will the reconciler model impact this in some way? Should we make some considerations in the way the `Update` logic is implemented here?

joamaki replied:
`Update` here essentially works at the same level of abstraction as our current code (https://github.com/cilium/cilium/blob/main/pkg/maps/lbmap/lbmap.go#L66) so same caviats apply here. I would make sure the same logic is implemented for backend ordering etc.

joestringer commented on pkg/loadbalancer/experimental/reflector.go:
```diff
@@ -0,0 +1,361 @@
+						// NOTE: Opting to panic on these failures for now to catch issues early.
+						// The production version of this needs to handle potential validation or
+						// conflict issues correctly.
+						panic("BUG: Upsert failed")
```
Out of curiousity, is there something about this code structure that makes handling the error difficult? It seems like these errors are currently out-of-scope for the initial experimental version here, but is that because we know how to improve this and want to focus efforts elsewhere on unknown areas, or is it because error handling here will be awkward and we don't yet know how to solve it?

On second thought, per the comment above the `ReflectorCell` about avoiding multiple references to the same underlying resources, I guess this code is subject to change anyway so it's probably something we'll just revisit when we get further down this road.

joestringer replied:
I guess logically, we should have an expression of any given k8s object into the internal form, but if for some reason we really can't convert it then probably the best we can do is to ignore this update and flag the resource somehow as unreconcilable. And yeah, doing that as early as possible would be ideal.

joestringer replied:
Ah, incidentally looking deeper into the error conditions at least for this particular call, it seems like they're primarily coding errors where the upsert is not done with proper ettiquette around transactions/locking.

joamaki replied:
This panic needs to be properly handled in the production code when `Services` actually does validation of the Upsert. For now it's just returning the errors from StateDB which indeed are related to misuse of WriteTxn (e.g. implementation bugs). 

It's still a bit of an open question on how we'd handle validation errors coming from e.g. `corev1.Service`. This could happen for example if we want to deny overriding of services if the source doesn't match. Should we keep retrying or should we drop it and, as you said, flag that it's invalid? Since we're still experimenting I opted to just panic to catch this immediately.

Another option for design is to e.g. accept a service from a different source but keep it as a "shadow" that will only become active when the primary goes away. In other words, try to design the `Services` API to be very lenient to what it accepts in order to not require retrying capabilities from "data sources" like this reflector. I don't see that working in the general case though as we do want to flag and report the issues towards the user.

joestringer commented on pkg/loadbalancer/experimental/reflector.go:
```diff
@@ -0,0 +1,361 @@
+				case resource.Delete:
+					name := loadbalancer.ServiceName{Namespace: obj.Namespace, Name: obj.Name}
+					if err := svcs.DeleteServiceAndFrontends(txn, name); err != nil {
+						panic("BUG: Delete failed")
```
Even if we're keeping `panic()` for the experimental version, the panic should probably report the underlying `err` and maybe an object name / frontend/ backend or some other info for further debugging...? (Likewise elsewhere)

joamaki replied:
Yes agreed, I'll fix these.

joestringer commented on pkg/loadbalancer/experimental/fuzz_test.go:
Just a heads up, I believe we only run these fuzz tests privately and get updates via the security list (cc @cilium/security-org). We can relay any updates we see there to you though, so I don't think it's a problem. If you want these to run regularly somewhere public then that might be a net-new addition to CI.

joamaki replied:
That's fine by me.

joestringer commented on pkg/loadbalancer/experimental/config.go:
```diff
@@ -0,0 +1,19 @@
+}
+
+func (def Config) Flags(flags *pflag.FlagSet) {
+	flags.Bool("enable-experimental-services", def.EnableExperimentalServices, "Enable use of the new experimental load-balancing API")
```
When you refer to API here, I assume you mean an unstable internal-only API, there's no current plans to commit to stable API for any of this?

joamaki replied:
I'll drop the "API" as it's more than that. And no, there's no plans for a stable API.

joestringer reviewed (approved):
The summary seems like a reasonable motivation + design. I can approve on behalf of @cilium/docs-structure. I'd like to see feedback from @cilium/sig-lb as owners of this code, they could speak best to the proposal.

Given that this code is experimental, I presume we'll need to go back through this code carefully again at a later date to ensure we've removed aspects we consider unsafe (eg `panic()`), added missing functionality, etc. I'll defer to @joamaki / @brb on how you want to manage that process as this implementation matures.

It's worth noting that we're on the tail end of the v1.16 development cycle, so folks may be more focused on hardening or other efforts rather than new functionality. Bear with us on the slow review cycle at this time. I'll welcome further reviews, though we should probably hold off on actually merging until after v1.16 is branched off (ETA next week).

brb reviewed (approved):
Nice! After merging, let's enable it in one of the ci-e2e configurations. `panic()` are fine for now, and they can be helpful to catch some errors.

A bigger discussion is needed how to move existing functionality to the new approach (LRP, L7, HC).

marseel commented on pkg/clustermesh/types/addressing.go:
```diff
@@ -42,6 +44,42 @@ type AddrCluster struct {
+	a.clusterID = a2.clusterID
+	return nil
+}
+
```
Could you add tests for it, please?

Simple `Unmarshal(Marshal(address))` == address with a few cases.

joamaki replied:
Added.

marseel commented on pkg/clustermesh/types/addressing.go:
```diff
@@ -42,6 +44,42 @@ type AddrCluster struct {
+	var b bytes.Buffer
+	b.WriteByte('"')
+	if !a.addr.IsValid() {
+		b.WriteString("0.0.0.0")
```
I'm not sure if it's a good idea to convert non valid address to a valid address, maybe just return error in that case instead?

joamaki replied:
Hmm the issue here is that the JSON marshalling is used for inspecting the data from cilium-dbg and if we error out on e.g. an uninitialized AddrCluster then we cannot inspect that. It also may be an OK situation to have it uninitialized and non-valid if the address isn't really used for anything in some configuration.

But yeah, turning this into "0.0.0.0" isn't correct either. I'll fix this up so that even for invalid addresses we preserve `Unmarshal(Marshal(data)) == data`.

marseel commented on pkg/clustermesh/types/addressing.go:
```diff
@@ -42,6 +44,42 @@ type AddrCluster struct {
+		return errors.New("AddrCluster.UnmarshalJSON: bad address")
+	}
+	// Drop the parens
+	data = data[1 : len(data)-1]
```
For `len(data) == 1` it will fail here with runtime error. I would probably switch check above from `len(data) == 0` to `len(data) <= 2`

marseel reviewed (changes requested):
Just reviewed clustermesh stuff, without bigger context for now.

marseel reviewed (approved):
Thank you, looks great for clustermesh.
